### DISCUSSION TITLE
Speaker Encoder questions (for VITS voice conversion)

### DISCUSSION THREAD

**Discussion by p0p4k** [coqui-ai/TTS#1480] [General Q&A] [answered]
Hello,
I am new to TTS. Loving the results of VITS model so far for a single speaker.
However, when training for a multi speaker dataset, the trainer generated a config.json and speakers.json file. While the training is still on-going, I wanted to ask how use the trained model for voice_conversion? Will there be a speaker encoder model saved separately? If not, then how do I get new speaker embedding using the VITS model? 
I used the parameter "use_speaker_embedding=True" in the config for training the VITS.
Please help, Thanks.

**Comment by erogol** [comment, marked as answer]
```use_speaker_embedding=True``` gets the model using the speaker embedding layer instead of the speaker encoder. You need to set the fields for the d_vector. 

You need to compute speaker d_vectors from the speaker encoder before you start training.  Use ```bin/compute_embeddings.py``` for this.

It generates a d_vector file. Then you pass this file to the training model, in your case VITS, by setting ```d_vector_file``` field in the config. 

Then enjoy your training. 

Note: we've not released the speaker encoder model separately. You need to pull it out of the released YourTTS model. Sorry, for the inconvenience at that part for now. 

@Edresson am I missing something?

**Reply by Edresson** [reply]
@erogol You are right.

 In addition, in the dev branch currently is possible to do voice conversion inference using the "tts" command line. Example:

`tts --model_path model.pth.tar --config_path config.json --encoder_path  model_se.pth.tar --encoder_config_path config_se.json --speakers_file_path speakers.json  --language_idx "en" --speaker_wav target_speaker.wav --reference_wav ref_speech.wav --out_path out_audio.wav`

**Reply by p0p4k** [reply]
@erogol @Edresson  
Thanks for your replies. I am learning so many things from this repo (python, AI TTS, etc). Really grateful.
1) I created d_vector_file using speaker encoder from YourTTS repo. 
2) Used umap-visualising jupyter notebook and plotted the graph, looks pretty good (3 speakers separate blobs plotted well). Just wanted to point out that the d_vector_files are now stored in .json format instead of .npx files separately (the umap notebook expects them in .npx), so maybe we can fix that minor stuff for next release. (i can try to PR if you okay with it)
3) Next, I will train VITS now for multispeaker, (i didnt pass the arguments to vitsargs earlier, only to vitsconfig), then will report the status.
4) [unrelated noob question] Why do my trained models get stored as just .pth and not .pth.tar (there is no difference in loading them for inference, just curious)?
5) [regarding results so far] I am training on Korean language (female,one speaker, 1 hour data) and trained a model for 440k steps (batch_size=60) already with very good results. Used this same model to fine tune (440+30k steps) on 20 minutes (female, one speaker) data, worked well. Then used it for fine tunning (440+15k steps) on 10 minutes (male, one speaker) data and it works good too! So thanks for the good efforts you guys put into this repo again!!!! (I can PR a basic korean text cleaner if its okay.)
6) I will read more papers and understand more about TTS and keep posting here. Have a good day!

**Reply by phamkhactu** [reply]
Hi @p0p4k 

As the step 5 above. I am really interested in your result? Can you give me more detail to get result? I understand that first of all, you get base model TTS by training **1 hour data**  then freeze encode text model for next  training step  **on 10 minutes (male, one speaker) data**. Do I comprehend it right?

Thank you.

**Reply by p0p4k** [reply]
@phamkhactu yes. At that time, I didn't know it is called "fine-tuning".

**Reply by phamkhactu** [reply]
@p0p4k I've encountered with a problem that if I freeze encode text,  a new speaker a add more(for example **10 minutes** ) will  not keep  new speaker style when TTS. Have you faced with it? if yes, how do you solve it?

**Reply by p0p4k** [reply]
@phamkhactu I don't remember now. But I think you can train without freezing anything and try as well.

**Comment by kin0303** [comment]
Hi,

I'm trying to train with vits for multi-speaker, I got an error like this, do you know the cause? Is it because of the dataset formatter?
```
--> STEP: 170/495 -- GLOBAL_STEP: 170
| > loss_disc: 2.56896 (2.49092)
| > loss_disc_real_0: 0.29256 (0.19180)
| > loss_disc_real_1: 0.21851 (0.25452)
| > loss_disc_real_2: 0.24393 (0.25421)
| > loss_disc_real_3: 0.27644 (0.25253)
| > loss_disc_real_4: 0.16534 (0.25395)
| > loss_disc_real_5: 0.19653 (0.23969)
| > loss_0: 2.56896 (2.49092)
| > grad_norm_0: 11.30908 (7.44597)
| > loss_gen: 2.27789 (2.37348)
| > loss_kl: 2.34740 (7.49759)
| > loss_feat: 3.50568 (3.26735)
| > loss_mel: 44.98423 (52.07333)
| > loss_duration: 2.61673 (2.72957)
| > loss_1: 55.73194 (67.94135)
| > grad_norm_1: 97.12949 (221.20049)
| > current_lr_0: 0.00020
| > current_lr_1: 0.00020
| > step_time: 0.75400 (0.63732)
| > loader_time: 0.00350 (0.00343)

! Run is kept in /media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/CoquiTTS/TTS/Vits-August-23-2022_02+00PM-68cef28a
Traceback (most recent call last):
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1492, in fit
self._fit()
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1476, in _fit
self.train_epoch()
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1254, in train_epoch
for cur_step, batch in enumerate(self.train_loader):
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 530, in next
data = self._next_data()
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1224, in _next_data
return self._process_data(data)
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1250, in _process_data
data.reraise()
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/_utils.py", line 457, in reraise
raise exception
RuntimeError: Caught RuntimeError in DataLoader worker process 3.
Original Traceback (most recent call last):
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
data = fetcher.fetch(index)
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/coqui_env/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 52, in fetch
return self.collate_fn(data)
File "/media/48e7e847-4b76-4839-a9bc-6f2511e980d3/riset_tts/Coqui_TTS/TTS/TTS/tts/models/vits.py", line 293, in collate_fn
wav_padded[i, :, : wav.size(1)] = torch.FloatTensor(wav)
RuntimeError: The expanded size of the tensor (1) must match the existing size (2) at non-singleton dimension 0. Target sizes: [1, 90736]. Tensor sizes: [2, 90736]
```

and it's my train.py
```
import os

from trainer import Trainer, TrainerArgs

from TTS.config.shared_configs import BaseAudioConfig
from TTS.tts.configs.shared_configs import BaseDatasetConfig
from TTS.tts.configs.vits_config import VitsConfig
from TTS.tts.datasets import load_tts_samples
from TTS.tts.models.vits import Vits, VitsArgs, CharactersConfig
from TTS.tts.utils.speakers import SpeakerManager
from TTS.tts.utils.text.tokenizer import TTSTokenizer
from TTS.utils.audio import AudioProcessor

output_path = os.path.dirname(os.path.abspath(file))
dataset_config = BaseDatasetConfig(name="mailab", meta_file_train="", path=os.path.join(output_path, ".../MAILAB/"))

audio_config = BaseAudioConfig(
sample_rate=22050,
win_length=1024,
hop_length=256,
num_mels=80,
preemphasis=0.0,
ref_level_db=20,
log_func="np.log",
do_trim_silence=True,
trim_db=23.0,
mel_fmin=0,
mel_fmax=None,
spec_gain=1.0,
signal_norm=False,
do_amp_to_db_linear=False,
resample=True,
)

vitsArgs = VitsArgs(
use_speaker_embedding=True,
)

config = VitsConfig(
model_args=vitsArgs,
audio=audio_config,
run_name="Vits_MAILAB",
num_speakers=4,
batch_size=8,
eval_batch_size=4,
batch_group_size=5,
num_loader_workers=4,
num_eval_loader_workers=4,
run_eval=True,
test_delay_epochs=-1,
epochs=2000,
text_cleaner="phoneme_cleaners",
use_phonemes=True,
phoneme_language="en-us",
phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
compute_input_seq_cache=True,
print_step=10,
print_eval=True,
mixed_precision=False,
output_path=output_path,
datasets=[dataset_config],
lr=0.0001,
test_sentences=["Aku lagi males.", "Males banget untuk kerja.", "Apa yang kamu lakukan jika kamu memiliki banyak uang?"],
)

ap = AudioProcessor.init_from_config(config)

tokenizer, config = TTSTokenizer.init_from_config(config)

train_samples, eval_samples = load_tts_samples(
dataset_config,
eval_split=True,
eval_split_max_size=config.eval_split_max_size,
eval_split_size=config.eval_split_size,
)

speaker_manager = SpeakerManager()
speaker_manager.set_ids_from_data(train_samples + eval_samples, parse_key="speaker_name")
config.model_args.num_speakers = speaker_manager.num_speakers

model = Vits(config, ap, tokenizer, speaker_manager)

trainer = Trainer(
TrainerArgs(),
config,
output_path,
model=model,
train_samples=train_samples,
eval_samples=eval_samples,
)
trainer.fit()
```

**Reply by manmaynakhashi** [reply]
@blackmamba1122 convert your wav file to mono.
